### PR TITLE
Rhmap 11585 list collections

### PR DESCRIPTION
**Pull request by wei-lee**
## Motive

Need a new endpoint to list the collections of the db

## Changes

* add a new collectionsList endpoint
* add integration tests

## JIRA

https://issues.jboss.org/browse/RHMAP-11585

### CODE REVIEW

File: src/middleware/db-connection/index.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+//this is for testing only. Should be replaced once the real code is ready
```
Comment (wei-lee) [inline src/middleware/db-connection/index.js:1]: @rachael-oregan  this is only for get integration tests work. Feel free to remove this one and replace it with the middleware you are working on.

File: src/endpoints/http/collections/index.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  //list collection info
+  router.get('/collections', function(req, res, next) {
+    var appname = req.param('appname');
+    req.log.debug({app: appname}, 'listing collections for app');
+    listCollections(req.param('appname'), req.log, req.db)
+      .then(result => {
+        req.log.trace({app: appname, result}, 'collection data listed');
+        res.json(result);
+      })
+      .catch(err =>  next(err));
```
Comment (pb82) [inline src/endpoints/http/collections/index.js]: why not just `catch(next)`?

File: src/endpoints/http/collections/list.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+ * Get the collection list and their info (including the namespace, name, count and size). Returns a promise.
+ * @param {string} appname The appname. It should be in the format of "<domain>-<appGuid>-<envId>".
+ * @param {object} logger The logger object
+ * @param {db} the db connection
+ * @returns Promise
+ */
+export default function listCollections(appname, logger, db) {
+  logger.debug({appname: appname}, "list collections for app");
+  return db.listCollections().toArray().then(items => {
+    const promises = items.filter(item => item.name.indexOf('system.') === -1).map(item => db.collection(item.name).stats());
```
Comment (pb82) [inline src/endpoints/http/collections/list.js]: Do we want to filter out the `systems` collections here? If its a shared db then fh-db should already do this. If it's db per app then why not show those collections to the user?

File: src/middleware/db-connection/index.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      if (err) {
+        return next(err);
+      }
+      fhdb.createMongoCompatApi({
+        __dbperapp: true,
+        connectionUrl: result.db
+      }).then(db => {
+        req.db = db;
+        next();
+      }).catch(err => {
```
Comment (pb82) [inline src/middleware/db-connection/index.js]: Can just be written as: `catch(next)`

File: src/app.js [hunk 5]
```diff
@@ -153,12 +151,45 @@ function main() {
...
     /* eslint-enable no-console */
-    process.exit(1);
+    return cb(err);
+  });
+}
+
+/**
+ * Stop the server. Used for testing purpose.
+ */
+export function stopServer(cb = EMPTY_FUNC) {
```
Comment (pb82) [inline src/app.js:170]: This is a useful concept. We should use default args more.

File: src/app.js [hunk 6]
```diff
@@ -133,15 +132,14 @@ function startApp(logger, fhconfig) {
...
   setupConfig()
   .then(config => {
     const logger = setupLogger(fhconfig);
     return {logger, config};
   })
   .then(param => {
-    if (args.d === true || args["master-only"] === true) {
+    if (args.d === true || args["master-only"] === true || process.env.NODE_ENV === 'test') {
       /* eslint-disable no-console */
       console.log("starting single master process");
```
Comment (pb82) [inline src/app.js]: The logger is already available here (param.logger) so why use `console.log`?

**Comment by wei-lee** [comment]
Retest this please

**Review by pb82 — COMMENTED** [review]
LGTM, just some comments

File: test/integration/setup.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
+/**
+ * Global Setup & Teardown functions for the integration tests
+ */
+var path = require('path');
```
Comment (sedroche) [inline test/integration/setup.js]: I was wondering why you revert to using `require` rather than `import` in these integration modules?

**Review by sedroche — COMMENTED** [review]
LGTM

File: src/endpoints/http/collections/index.js (hunk 2 above)
Comment (wei-lee) [reply src/endpoints/http/collections/index.js]: yes, that's right. Will fix it

File: src/endpoints/http/collections/list.js (hunk 3 above)
Comment (wei-lee) [reply src/endpoints/http/collections/list.js]: that's true, we can show them to users. Not sure if there will be any side effects, but we can try it.

File: src/middleware/db-connection/index.js (hunk 4 above)
Comment (wei-lee) [reply src/middleware/db-connection/index.js]: yes

File: src/app.js (hunk 6 above)
Comment (wei-lee) [reply src/app.js]: ah, yes. will replace it with logger

File: test/integration/setup.js (hunk 7 above)
Comment (wei-lee) [reply test/integration/setup.js]: yeah, now I am asking the same question to myself ;). I think I will revert it back to `import`

File: api-docs/swagger.yaml [hunk 8]
```diff
@@ -0,0 +1,127 @@
+swagger: '2.0'
```
Comment (wei-lee) [inline api-docs/swagger.yaml:1]: @sedroche  I convert the swagger.json file to yaml file as yaml file is a lot easier to read and write. Mind reviewing this?

File: src/endpoints/http/collections/index.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+    req.log.debug({app: appname}, 'listing collections for app');
+    listCollections(req.param('appname'), req.log, req.db)
+      .then(result => {
+        req.log.trace({app: appname, result}, 'collection data listed');
+        res.json(result);
+      })
+      .catch(next);
+  });
+  //create collection
+  createCollection(router);
```
Comment (wei-lee) [inline src/endpoints/http/collections/index.js:17]: @sedroche  I moved your collection create code to the `collections` dir to keep the code together, and made some changes. are you ok with the change?

File: api-docs/swagger.yaml (hunk 8 above)
Comment (sedroche) [reply api-docs/swagger.yaml:1]: Yep, I don't mind either way, but this looks good to me @wei-lee

File: src/endpoints/http/collections/index.js (hunk 9 above)
Comment (sedroche) [reply src/endpoints/http/collections/index.js:17]: Yep, makes sense. Changes look fine to me 👍

**Pull request merged**